### ISSUE TITLE
2016-04-27 Outage (7:10AM-7:18AM)

### ISSUE THREAD

**Issue by GUI** [18F/api.data.gov#333]
We experienced an ~8 minute outage from 7:10AM-7:18AM ET that affected all agencies using the api.data.gov service.

The root cause was that the log disks on our proxy servers filled up, which then started to make requests fail. A couple of additional notes:

- We have alerting setup to notify us when our disks are approaching full, but these alerts didn't go out soon enough. We received alerts at 3:43AM and 4:12AM ET.
- We experienced a bit of an abnormal traffic surge starting last night, which contributed to the disks filling up faster than our normal usage patterns.

We freed up some disk space on the log partition to get things back in action quickly, but in terms of really remediating the issue, here are some next steps:

- Expand the log disk partitions so we have more available space to grow.
- Adjust our alerting so we get notified about these kind of issues earlier.
- Fix things so that even if the log disk fills up, the service should remain operational. On the proxy servers, we don't want the log partition being full to result in these kind of fatal error. I think nginx should be able to operate with a full log disk (which is one of the primary reasons we partitioned it separately to begin with), but need to investigate more about which specific component was failing.
- Look into setting up auto-scaling. We now have more of the necessary pieces in place (CloudFormation templates for our servers), so auto-scaling may also be something we want to explore to make the service more resilient to unexpected failures like this (since auto-scaling could have automatically spun up new servers to replace the bad ones).

**Comment by timwood** [comment]
What about remote logging to a splunk or similar logging server/service? Or are you already doing that?

**Comment by GUI** [comment]
That's an excellent point as well. We are already logging request data to a centralized database (for analytics purposes), but we still do log a variety of things to local files as well (partially since it's the default, and partially for redundancy and debugging). However, revisiting our overall approach to logging for this other data is also a good idea. Thanks!

**Comment by GUI** [comment]
The first two items have been tackled to address this, so we have significantly more buffer space, and we get alerted much earlier about any potential disk space issues.

I've opened separate stories for centralized logging (https://github.com/18F/api.data.gov/issues/337) and auto-scaling (https://github.com/18F/api.data.gov/issues/336). Those are larger issues we may consider in the future, but since the basic issue has been addressed, I'm going to go ahead and close this issue.
